### PR TITLE
Enable custom configuration settings for the data table manager

### PR DESCRIPTION
**Pull request by ddouglasz**
#### Summary

Enable custom configuration settings for data table amanager

## Description
We want to be able to allow users to create their desired table configurations using the data table manager. This will allow granular configurations for the users as they can now create more settings, beyond the currently existing columns settings and display settings.

Reference: https://github.com/commercetools/ui-kit/issues/2687#issuecomment-2141801498

### CODE REVIEW

File: packages/components/data-table-manager/src/data-table-manager.story.js [hunk 1]
```diff
@@ -293,6 +300,122 @@ storiesOf('Components|DataTable', module)
...
+              </Spacings.Stack>
+            </Grid.Item>
+          </Grid>
+        </>
+      );
+    };
+
+    const customSettings = [
+      {
+        id: 'customSettings1',
```
Comment (ddouglasz) [inline packages/components/data-table-manager/src/data-table-manager.story.js:396]: The payload can take as much settings as the user wants to access in the nested rows.

File: packages/components/data-table-manager/src/data-table-manager.story.js [hunk 2]
```diff
@@ -350,6 +477,8 @@ storiesOf('Components|DataTable', module)
       'showColumnManagerConfirmationButtons',
       false
     );
+    // TODO - remove when nexted rows are implemented
+    const debug = boolean('logCustomSettings', false);
```
Comment (ddouglasz) [inline packages/components/data-table-manager/src/data-table-manager.story.js:481]: This is solely for storybook and testing purpose. Will be helpful when implementing the nested rows. It is currently a toggle in storybook to help see the data passed in the customSettingsPayload

**Comment by jaikumar-tj** [comment]
@ddouglasz Thanks for doing this. I took a quick glance at the code and noticed a couple of things that might be crucical for adapting in MC products app. Please let me know your thoughts or any alternative suggestion to resolve this.

1. The `customSettings` are appended as the last options inside `getDropdownOptions` function. Would it be possible to inject the `customSettings` based on desired `position` ? For instance, we want to keep `product columns` and `variant column` right next to each other.
2. We would like to provide custom label for the existing option `column manager` in order to achieve our desired outcome as shown in the screenshot. Do you think this would be possible? Otherwise the wording might looks odd if one option is `column manager` and another is `variant columns`

 
Screenshot 2024-07-11 at 14 21 54

**Comment by ddouglasz** [comment]
Thanks @jaikumar-tj for the feedback and suggestions. To answer your questions:
> 1. The `customSettings` are appended as the last options inside `getDropdownOptions` function. Would it be possible to inject the `customSettings` based on desired `position` ? For instance, we want to keep `product columns` and `variant column` right next to each other.

Good point, I don't think that would be too hard to implement. I will look into it. 


> 2. We would like to provide custom label for the existing option `column manager` in order to achieve our desired outcome as shown in the screenshot. Do you think this would be possible? Otherwise the wording might looks odd if one option is `column manager` and another is `variant columns`
> 

This makes sense, since now we have to identify every layer of the custom rows we are controlling. Maybe a small caveat I see is backwards compatibility, but likely can still work fine by keeping the initial default behaviour when no custom label is provided.

**Comment by ddouglasz** [comment]
> The variant columns option re-uses the same ColumnSettingsManager. If I am not wrong, the ColumnSettingsManager is not available outside of UI-kit. In that case, can we make it available or should the Pacman team create a custom component (I think this might lead to the same problem of deviating from UI-kit) ?

I think this is the part of the purpose of the feature, the way I see it, you have a blank canvas to put any component of choice so I do not think creating your own column manager would affect maintainability in any way since that was mainly one of the issues. 
Whatever update is made on uikit, that settings interface will always be there to support whatever change you make to your custom component (be it a column manager). An alternative is to consider giving the option of passing either a custom component or a column manager. I would like to get the opinions of the rest of the @commercetools/shield-team-design-system  team too

**Comment by jaikumar-tj** [comment]
@ddouglasz I understand your point about "blank canvas" that supports any component of choice. 

> so I do not think creating your own column manager would affect maintainability in any way since that was mainly one of the issues.

I wanted to provide some context and rephrase my concern. If the Pacman team creates its own column manager, we might need to copy/paste the [column-setting-manager](https://github.com/commercetools/ui-kit/tree/main/packages/components/data-table-manager/src/column-settings-manager) from the UI-kit repo into the MC-products apps, as we want the same exact component for the `variant column manager`. @tdeekens's [initial concern](https://commercetools.slack.com/archives/C073XK3H99Q/p1715939428868449) was to avoid design drift and ensure convergence with UI-kit, preventing any friction or delays during the UI-kit's rollout.

If we opt for a custom column manager, any new updates from UI-kit will affect the `product column manager` but not the `variant column manager`. **For background context**: When we (Pacman team) diverged from UI-kit, we accepted the tradeoff of not being visually consistent with the rest of MC. We were okay with the `<DataTableManager>` in other parts of MC having new features or changes, and we didn’t expect to adapt immediately. However, from Tobi's comment, it seems this has caused some friction with some UI-kit release PRs.

To rephrase my concern: In the future, If there is a new change (for example, a background color update) in the `<ColumnSettingManager>` and a new version of UI-kit is released, will the team owning the UI-kit face any friction or delays in rolling out the UI-kit because the `variant column manager` is using a copy/pasted custom column manager?

Screenshot 2024-07-11 at 16 22 51

**Comment by ddouglasz** [comment]
Thank you for the elaborate explanation 🙏🏾. 
We can add it as an optional property that can be provided or not in the custom settings.

**Comment by ddouglasz** [comment]
For the feedback with changing the default settings label and changing the order in which they come has been addressed here: https://github.com/commercetools/ui-kit/pull/2851/commits/d20315230117945b79a26fce0a04d02da5902ca1

Thank you.

**Review by jaikumar-tj — APPROVED** [review]
So far the code and the implementation looks fine 👍 

> We can add it as an optional property that can be provided or not in the custom settings.

Just to be clear, when you say `optional property`, do you mean to introduce a new prop?

I was wondering since the current "blank canvas" approach is fine and  if the `ColumnSettingsManager` is exported and available outside UI-kit then I can simply pass that component to `customComponent` prop as below

```jsx
 const customSettings = [
      {
        title: 'Variant columns',
        value: 'variant-columns',
        customComponent: <ColumnSettingsManager ... />,
```

Do you think this is very complex to achieve? @ddouglasz If this comment is causing confusion, I'm happy to have a short call to align ourselves.

**Comment by ddouglasz** [comment]
>I was wondering since the current "blank canvas" approach is fine and if the ColumnSettingsManager is exported and available outside UI-kit then I can simply pass that component to customComponent prop as below

Ah, I see what you mean now. Yes, we can try this. Let's see how complex it is to expose the ColumnSettingsManager. It's a great idea. Thank you.

**Review by obulaworld — APPROVED** [review]

**Review by jocelyn-dunkley — APPROVED** [review]

**Review by misama-ct — APPROVED** [review]
🎉  Great work! Happy to see that so many others approve as well. Thank you for tackling this 🙇‍♂️

**Review by FilPob — APPROVED** [review]

File: packages/components/data-table-manager/README.md [hunk 3]
```diff
@@ -112,40 +112,54 @@ export default Example;
...
+| `customColumnManager.visibleColumnKeys`              | Array of `string`                                           |          |         | The keys of the visible columns.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   |
```
[the remark this answers is not in the document]
Comment (ddouglasz) [reply packages/components/data-table-manager/README.md:156]: Thank you @jaikumar-tj ,
Nice catch! I have updated the readme [here](https://github.com/commercetools/ui-kit/pull/2851/commits/0f672549f9315930e5ac8b34d061b77f904f891a) to describe it's use. 
I missed the proper cleanup of the property. Remember from our last feedback, we also want to have the option of overriding the default manager labels. When this value is provided, it overrides the label of the default manager labels.

[the remark this answers is not in the document]
Comment (jaikumar-tj) [reply packages/components/data-table-manager/README.md:156]: @ddouglasz 
Thanks for updating readme. Here is my overall understanding and let me know if it is correct

<table>
  <tr>
    <td> `columnManager.columnManagerLabel`</td>
    <td>
Screenshot 2024-08-01 at 08 58 03

</td>
   </tr> 
   <tr>
      <td>`customSettings['key']. customPanelTitle`</td>
      <td>
Screenshot 2024-08-01 at 08 57 22
  </td>
  </tr>
  <tr>
      <td>`customSettings['key'].customComponent`</td>
      <td>
Screenshot 2024-08-01 at 08 57 50
  </td>
  </tr>
    </tr>
  <tr>
      <td>`customColumnManager`<li>customColumnManager.hideableColumns</li>
      <li>customColumnManager.visibleColumnKeys </li>
            <li>customColumnManager.disableColumnManager  ... etc</li>
      </td>
      <td>
???
  </td>
  </tr>
</table>

What about `customColumnManager`? When do we have to use that? You mentioned that "It is only meant for the already existing data table managers" but I'm not sure if you are referring to this prop or the one about overriding label.

[the remark this answers is not in the document]
Comment (jaikumar-tj) [reply packages/components/data-table-manager/README.md]: @ddouglasz Feel free to merge the PR. I think my doubts might be clarified once we start implementing this new changes in the product list.

[the remark this answers is not in the document]
Comment (ddouglasz) [reply packages/components/data-table-manager/README.md]: Sure! I pushed the change from our meeting yesterday. Thanks for the review. Will be happy to take align with you at any point, if you have more questions.
https://github.com/commercetools/ui-kit/pull/2851/commits/6af7dcfccdd35dc44c89a7b332bde646cff932e9

Thank you.

**Pull request merged**